### PR TITLE
Automatically retrying on 503 errors

### PR DESCRIPTION
**Pull request by hiranya911**
* Increased retry count to 4
* Backoff factor set to 0.5 (this results in the delay pattern of 0s, 1s, 2s and 4s)

### CODE REVIEW

File: test/unit/utils/api-request.spec.ts [hunk 1]
```diff
@@ -623,56 +623,91 @@ describe('HttpClient', () => {
     }).should.eventually.be.rejectedWith(err).and.have.property('code', 'app/network-error');
   });
 
-  it('should timeout when the response is delayed', () => {
+  it('should timeout when the response is repeatedly delayed', () => {
     const respData = {foo: 'bar'};
     const scope = nock('https://' + mockHost)
       .get(mockPath)
-      .twice()
+      .times(5)
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: I think it is better to test passing explicit RetryConfig instead so you don't have to do things like:
`(client as any).retry.backOffFactor = 0;`

This also makes it easier to add changes to default config in the future.

And then add a test for default config.

Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Done

File: test/unit/utils/api-request.spec.ts [hunk 2]
```diff
@@ -766,11 +815,7 @@ describe('HttpClient', () => {
         'content-type': 'application/json',
       });
     mockedRequests.push(scope2);
-    const client = new HttpClient({
-      maxRetries: 1,
-      maxDelayInMillis: 1000,
-      statusCodes: [503],
-    });
+    const client = new HttpClient();
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: Shouldn't this use `testRetryConfig()` ?

File: test/unit/utils/api-request.spec.ts [hunk 3]
```diff
@@ -847,13 +888,9 @@ describe('HttpClient', () => {
         'content-type': 'application/json',
       });
     mockedRequests.push(scope);
-    const client = new HttpClient({
-      maxRetries: 4,
-      backOffFactor: 0.5,
-      maxDelayInMillis: 60 * 1000,
-      statusCodes: [503],
-    });
+    const client = new HttpClient();
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: Should you pass explicit test retry config with backoff factor on, here and elsewhere?

File: test/unit/utils/api-request.spec.ts [hunk 4]
```diff
@@ -92,6 +92,18 @@ function mockRequestWithError(err: any) {
     .replyWithError(err);
 }
 
+/**
+ * Returns a new RetryConfig instance for testing. This is same as the default
+ * RetryConfig, with the backOffFactor set to 0 to avoid delays.
+ *
+ * @return {RetryConfig} A new RetryConfig instance.
+ */
+function testRetryConfig(): RetryConfig {
```
Comment (bojeil-google) [inline test/unit/utils/api-request.spec.ts]: should this take a parameter whether to reset backOffFactor and then you can use it in all tests not testing default behavior but require backOffFactor?

File: test/unit/utils/api-request.spec.ts (hunk 2 above)
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Done

File: test/unit/utils/api-request.spec.ts (hunk 4 above)
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: That's same as default config. So using `defaultRetryConfig()` directly.

File: test/unit/utils/api-request.spec.ts (hunk 3 above)
Comment (hiranya911) [reply test/unit/utils/api-request.spec.ts]: Used `defaultRetryConfig()` directly.

**Comment by hiranya911** [comment]
@bojeil-google I used `defaultRetryConfig()` explicitly in the tests you pointed out to make the intention clear. Please take a look.

**Pull request merged**